2319: handle withdrawn applications

andrew8er opened the pull request:
### Short Description

Enables correct handling of withdrawn applications. This includes administration frontend and the backend. This was a little bit more involved than initially estimated, since to correctly handle the filtering and sorting, we need to take the application's pre-verification status into account, which is hidden in the raw `jsonValue` string data. There were four different places, where this property was parsed and passed around separately from the application object.

The `jsonValue` property is now parsed (in `parseApplication()` immediately after receiving the application objects from the different queries, which simplifies the passing of data a bit.

Please have a close look at the backend part of the PR, I might have missed a spot where a withdrawn application should have been handled.

### Proposed Changes

#### Major changes

- Remove DB column `application.withdrawalDate` and replace it with `statusResolvedDate`. The withdrawal is already captured with the `status` column.
- Removed the `withdrawalDate` from the GraphQL type, replace with `statusResolvedDate`.
- Show the withdrawal notice in the `<ApplicationCard>` component.
- Preparse the `application.jsonValue` immediately after querying and pass it down to components.
- Respect the pre-verification status when sorting and filtering applications in `<ApplicationOverview>`.
- In the backend, prohibit some mutations on applications when an application is withdrawn.

#### Minor changes

- Improve the IntelliJ SQL code style, it is now a little less horrible.
- Add Freinet demo administrator and application.
- Add demo data for an undecided application

### Side Effects

<!-- List all related components that have not been explicitly changed but may be affected by this PR -->

### Testing

- Inspect the filtering and sorting in http://localhost:3000/applications

### Resolved Issues

Fixes: #2319

f1sh1918 commented on administration/src/mui-modules/application-verification/ApplicationApplicantController.tsx:
```diff
@@ -23,9 +24,9 @@ const ApplicationApplicantController = ({ providedKey }: { providedKey: string }
     return applicationQueryHandler.component
   }
-  const application = applicationQueryHandler.data.application
+  const application = parseApplication(applicationQueryHandler.data.application)
```
❌ parsing the application here, leads to null values for some properties

<img width="396" height="115" alt="image" src="https://github.com/user-attachments/assets/53279b00-61ad-4dc9-b89a-1990e675036d" />

This also leads to `ApplicantApplicantView` will not receive proper status

f1sh1918 commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/application/webservice/schema/view/ApplicationView.kt:
```diff
@@ -25,7 +24,6 @@ data class ApplicationView(
                 regionId = entity.regionId.value,
                 createdDate = entity.createdDate.toString(),
                 jsonValue = entity.jsonValue,
-                withdrawalDate = entity.withdrawalDate?.toString(),
```
❌ Related to LN 29- LN 30
Even this issue came with an pr before, i just realized it while testing the `ApplicationApplicantView`

This information will not be passed for `getApplicationByApplicant` 
We need this information in the `ApplicationApplicantController`

f1sh1918 reviewed (changes requested):
Looks already really good but i have to request changes here, since the `ApplicationApplicantView` breaks or has been broken (before)
1. Create an application
2. Open the applicant status link
http://localhost:3000/antrag-einsehen/<accesskey>/
3. There should be a withdraw button

I added a comment why this happens, check `ApplicantView`
The return model has some privacy settings that prevent `status` and `statusResolveDate` to be exposed for non logged in users.

It would be also okay to fix it in a separate issue. But it just affects two lines of code

bahaaTuffaha reviewed (commented):
<img width="400" alt="image" src="https://github.com/user-attachments/assets/b8c980e4-ee58-45e0-bce9-300a6430e6c0" />

This is pending and I can't see the withdraw button.

f1sh1918 commented:
> <img alt="image" width="400" src="https://private-user-images.githubusercontent.com/101111168/472652106-b8c980e4-ee58-45e0-bce9-300a6430e6c0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM5NjcwMzUsIm5iZiI6MTc1Mzk2NjczNSwicGF0aCI6Ii8xMDExMTExNjgvNDcyNjUyMTA2LWI4Yzk4MGU0LWVlNTgtNDVlMC1iY2U5LTMwMGE2NDMwZTZjMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDczMVQxMjU4NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YzhlNGVhZTgzNmNjMzE4ZjM2MThjOTAwNDNiMjVlNTc0YjczYjBmMzFjNDdjNTYxNDEyMjQzMjNkMWEyMGZkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.J5ZYErLvY5OdI4MRffLvsXVT7Yeey23ZNoR4FDTWynE">
> This is pending and I can't see the withdraw button.

see why i requested changes the comment before

seluianova commented on administration/src/bp-modules/applications/ApplicationCard.tsx:
```diff
@@ -229,43 +228,69 @@ const ButtonsApplicationResolved = (props: {
+    case ApplicationStatus.Rejected:
+      return 'error'
+    case ApplicationStatus.Withdrawn:
+      return 'warning '
```
```suggestion
      return 'warning'
```

seluianova commented:
Broken translations:
<img width="881" height="416" alt="image" src="https://github.com/user-attachments/assets/d221dcd1-0b2d-4be3-bd99-ba2f27800ba9" />

For other statuses as well:
<img width="879" height="425" alt="image" src="https://github.com/user-attachments/assets/e7562958-43fe-4e23-a44b-876567d79baf" />

seluianova commented:
Something messed up with the counts
64 =! 6 + 2 + 2 + 56
<img width="970" height="59" alt="image" src="https://github.com/user-attachments/assets/7b481bb5-96a6-4614-a0a1-1dc9f2874baa" />

seluianova commented on administration/src/bp-modules/applications/ApplicationsOverview.tsx:
```diff
@@ -1,39 +1,65 @@
-  {
-    title: 'allApplications',
-    status: undefined,
+const countApprovingVerifications = (application: Application): number =>
```
countApprovedVerifications ?
countRejectedVerifications

seluianova commented on administration/src/bp-modules/applications/ApplicationsOverview.tsx:
```diff
@@ -1,39 +1,65 @@
+  accepted: {
+    i18nKey: 'accepted',
+    filter: (application: Application): boolean =>
+      countApprovingVerifications(application) > 0 ||
```
the application is approved when every verification is approved, not just > 0

seluianova commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/application/webservice/schema/view/ApplicationAdminGql.kt:
```diff
@@ -0,0 +1,38 @@
+import java.util.concurrent.CompletableFuture
+
+@GraphQLName("ApplicationAdmin")
+data class ApplicationAdminGql(
```
🙃 I think the previous name was clearer tbh

seluianova commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/application/webservice/EakApplicationMutationService.kt:
```diff
@@ -249,3 +259,11 @@ class EakApplicationMutationService {
+        this.status = status
+        this
+    } catch (e: IllegalArgumentException) {
+        throw InvalidInputException("Cannot approve application, status is '${this.status}'")
```
this function is used not only for approval but also for rejection

seluianova commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/application/webservice/schema/view/ApplicationView.kt:
```diff
@@ -1,53 +0,0 @@
-import java.util.concurrent.CompletableFuture
-
-@Suppress("unused")
-data class ApplicationView(
```
I didn't get why this data class was split up tbh.
I guess the idea was to hide some fields from the applicant. but this can be done at the graphql query level.
we can remove unnecessary fields from the query, and they will not be returned
```
query GetApplicationByApplicant {
    getApplicationByApplicant(accessKey: "ghjg") {
        createdDate
        id
        jsonValue
        regionId
        status
        statusResolvedDate
        withdrawalDate
    }
}
```

andrew8er replied:
Several reasons:
- The most important one was that it is conceptually clearer to have several "views" for different audiences and thus to avoid accidentally revealing data. As @f1sh1918 [pointed out here](https://github.com/digitalfabrik/entitlementcard/pull/2373#pullrequestreview-3063029355), there might even be a third view necessary. I tend to agree here. Having different view types at the lowest level (server controller layer) removes the possibility for accidents in any of the upper levels (e.g. UI).
- The other one is that having separate types removes some unnecessary "nullable" properties, which were only nullable, because they might need to be omitted for more restricted views. The nullability also trickles down to the client side and having properties non-nullable removes some redundant null-checks.

> I guess the idea was to hide some fields from the applicant. but this can be done at the graphql query level.

This would open an information disclosure vulnerability. The attributes a GraphQL query retrieves of an object is simply a client side construct meant to avoid transmitting unnecessary data and nothing would stop an attacker from requesting the "hidden" attributes if you do not prohibit this at the server.

seluianova commented on administration/src/shared/verifications.ts:
```diff
@@ -0,0 +1,21 @@
+import { ApplicationVerificationView } from '../generated/graphql'
+
+export const isEmailValid = (email: string): boolean => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)
```
🙃 this function is used not only for verifications but also to check applicant email (and can be potentially used to verify other emails, e.g. when creating a new admin), not sure if this function should be moved to "verifications.ts"

andrew8er replied:
To me this is a very generic utility function, I would place that in a file `/src/misc/emails.ts` or so.

What worries me more is the possibility that this regex will reject valid email addresses because it is hilariously simplified for what it claims to do. 

For the curious, a regex that correctly validates real world RFC822 email addresses looks like [this](https://pdw.ex-parrot.com/Mail-RFC822-Address.html), but we should use a specialized package, since it is really slow.

seluianova commented:
The button to re-send an approval request is not available if the application has multiple verifications and one of them has been approved.
<img width="1349" height="737" alt="image" src="https://github.com/user-attachments/assets/91dd904b-5c22-4116-bc63-a07b1386da47" />

seluianova reviewed (changes requested):
Adding “request changes” as I already mentioned some breaking issues, but I have not checked everything yet as it is really a lot.

andrew8er commented:
> Broken translations:
> For other statuses as well:

Good catch. I expanded the i18n a little bit and separated the status bar texts from the "no applications" texts for grammatical correctness.

> Something messed up with the counts / filters by status
> 64 =! 6 + 2 + 2 + 56
> 2 applications are lost somewhere in "all applications" and cannot be filtered

Not necessarily, since the different statuses are not strictly mutually exclusive. An application can be both "withdrawn" (from the `status` field) and be either approved/rejected/open (from the application verifications _or_ the preverification status).

We had a discussion [here](https://github.com/digitalfabrik/entitlementcard/issues/2319#issuecomment-3102446784), but it could be argued that withdrawn applications should be removed from all other filters.

> I think it’d be better to open separate PRs for large refactorings.
> It’s a pretty big change (1000+ lines), and seems to require an extra testing.

Generally, you're totally right. The thing here is, I did not plan to make this huge refactoring, I only noticed that it would be necessary pretty late in the process, when fine-tuning the sorting of the application list. I think this PR is better review on a commit by commit basis.

> btw the testing instructions don't include the key changes requested in the task (handling of the “withdrawn” status)

This is because I do not have a firm grasp of the ways the backend processes applications (especially with respect to sending emails). I tried to wade through the code, but it's pretty convoluted imho. So I added just some very general instructions and a note for the backend in the PR introductory text.

> The button to re-send an approval request is not available if the application has multiple verifications and one of them has been approved.

- Original logic was `application.status !== ApplicationStatus.Rejected`. 
- Current logic is `application.status === ApplicationStatus.Pending`, 
- Now changed to `verificationStatus(verification) === VerificationStatus.Pending`. Is this correct?

seluianova commented:
> Original logic was application.status !== ApplicationStatus.Rejected.
> Current logic is application.status === ApplicationStatus.Pending,
> Now changed to verificationStatus(verification) === VerificationStatus.Pending. Is this correct?

No 🙈
Original logic was:
Show button if **verification is pending** AND **application is not rejected**

Expected change within the [2319](https://github.com/digitalfabrik/entitlementcard/pull/2373/files#top):
Show button if **verification is pending** AND **application is not rejected** AND **application is not withdrawn**

andrew8er commented:
> Show button if verification is pending AND application is not rejected AND application is not withdrawn

👍  Updated and rebased.

Docmale commented:
found:
Withdrawn data are not deleted #2510
